Implement the Close Watcher API

lukewarlow opened the pull request:
#### 74e0b308c9978fbbd289a8b070f3352c411d173d
<pre>
Implement the Close Watcher API
<a href="https://bugs.webkit.org/show_bug.cgi?id=263305">https://bugs.webkit.org/show_bug.cgi?id=263305</a>

Reviewed by Abrar Rahman Protyasha.

Implement the Close Watcher API behind a feature flag.

This patch doesn&apos;t include the integration with Popovers or Dialogs.

Spec: <a href="https://html.spec.whatwg.org/#close-requests-and-close-watchers">https://html.spec.whatwg.org/#close-requests-and-close-watchers</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/abortsignal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/keydown-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/keypress-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/keyup-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/not-user-activation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/synthetic-keyboard-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/event-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/frame-removal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/inside-event-listeners-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n-activate-preventDefault_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n-activate_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n-closerequest-n_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n-destroy-n_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nn-activate-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nn-dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nnn-CloseWatcher-dialog-popover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nnn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/ny-activate-preventDefault_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/ny_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyn_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyn_dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nynn-destroy_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nynn-destroy_dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nynn_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nynn_dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyyn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyyyn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/y_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yn-activate_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yn_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/ynn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yy_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yyn_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yyy-CloseWatcher-dialog-popover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yyy-activate-CloseWatcher-dialog-popover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yyy_CloseWatcher-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/close-watcher/abortsignal-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/close-watcher/basic-expected.txt: Added.
* LayoutTests/platform/win/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/html/closewatcher/CloseWatcher.cpp: Added.
(WebCore::CloseWatcher::create):
(WebCore::CloseWatcher::establish):
(WebCore::CloseWatcher::CloseWatcher):
(WebCore::CloseWatcher::requestClose):
(WebCore::CloseWatcher::requestToClose):
(WebCore::CloseWatcher::close):
(WebCore::CloseWatcher::canBeClosed const):
(WebCore::CloseWatcher::destroy):
(WebCore::CloseWatcher::eventListenersDidChange):
(WebCore::CloseWatcher::virtualHasPendingActivity const):
* Source/WebCore/html/closewatcher/CloseWatcher.h: Added.
* Source/WebCore/html/closewatcher/CloseWatcher.idl: Added.
* Source/WebCore/html/closewatcher/CloseWatcherManager.cpp: Added.
(WebCore::CloseWatcherManager::add):
(WebCore::CloseWatcherManager::remove):
(WebCore::CloseWatcherManager::notifyAboutUserActivation):
(WebCore::CloseWatcherManager::canPreventClose):
(WebCore::CloseWatcherManager::escapeKeyHandler):
* Source/WebCore/html/closewatcher/CloseWatcherManager.h: Added.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::defaultKeyboardEventHandler):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::notifyActivated):
(WebCore::LocalDOMWindow::closeWatcherManager):
* Source/WebCore/page/LocalDOMWindow.h:

Canonical link: <a href="https://commits.webkit.org/289529@main">https://commits.webkit.org/289529@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e89939f183aef320b7c3387aeb101acef6c93759

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87191 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6701 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41551 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92052 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37931 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89241 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6977 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14770 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67397 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25127 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90193 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5344 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78918 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47717 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5117 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33297 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37047 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/80004 "Found 18 new JSC stress test failures: microbenchmarks/elidable-new-object-then-call.js.lockdown, microbenchmarks/map-iteration-and-array-destructuring.js.lockdown, microbenchmarks/simple-regexp-exec-folding.js.lockdown, microbenchmarks/todomvc-javascript-es6-webpack-json-parse.js.lockdown, microbenchmarks/varargs-inline.js.lockdown, stress/big-int-multiply-memory-stress.js.lockdown, stress/check-is-constant-non-cell-should-not-array-profile-during-osr-exit.js.lockdown, stress/generator-fib-ftl-and-array.js.lockdown, stress/multi-put-by-offset-multiple-transitions.js.lockdown, stress/object-assign-empty.js.lockdown ... (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75600 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34173 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93937 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/85965 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14353 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10441 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76196 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14557 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74774 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75399 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18555 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19738 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18175 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7282 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14372 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19665 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/108459 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14117 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26099 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17560 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15898 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/html/closewatcher/CloseWatcher.cpp:
```diff
@@ -0,0 +1,149 @@
+{
+    RefPtr document = dynamicDowncast<Document>(&context);
+    if (!document || !document->domWindow())
+        return Exception { ExceptionCode::InvalidStateError, "CloseWatchers cannot be created in a detached window."_s };
```
If a document does not have a domWindow can it ever be fully active? I think not so this check might be redundant? If it's not redundant, we are missing this check elsewhere.

lukewarlow replied:
Yep I think you're right, the tests all pass with this check removed so I've gone ahead and dropped it.

annevk reviewed (commented):
This looks reasonable to me, but this should have review from someone with more C++ familiarity as well.

Also note that WebKit's standards-position on this is not positive. It's neutral. This patch is still appreciated of course and I think we should take it provided everyone is on board with the code, but it's also important to not misrepresent the project's position on this feature.

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcher.cpp:
```diff
@@ -0,0 +1,149 @@
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(CloseWatcher);
```
I recognize you wrote this original patch a while back, but this is now a legacy macro and we should use TZone allocation instead.

i.e. `WTF_MAKE_TZONE_ALLOCATED_IMPL`

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcher.cpp:
```diff
@@ -0,0 +1,149 @@
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(CloseWatcher);
+
+ExceptionOr<Ref<CloseWatcher>> CloseWatcher::create(ScriptExecutionContext& context, const CloseWatcher::Options& options)
```
Nit, you can drop the class scope after the function name.
```suggestion
ExceptionOr<Ref<CloseWatcher>> CloseWatcher::create(ScriptExecutionContext& context, const Options& options)
```

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcher.cpp:
```diff
@@ -0,0 +1,149 @@
+
+ExceptionOr<Ref<CloseWatcher>> CloseWatcher::create(ScriptExecutionContext& context, const CloseWatcher::Options& options)
+{
+    RefPtr document = dynamicDowncast<Document>(&context);
```
Nit: no need for a pointer
```suggestion
    RefPtr document = dynamicDowncast<Document>(context);
```

aprotyas replied:
(we have overloads to cover both cases)

```cpp
inline match_constness_t<Source, Target>* dynamicDowncast(Source& source)

inline match_constness_t<Source, Target>* dynamicDowncast(Source* source)
```

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcher.cpp:
```diff
@@ -0,0 +1,149 @@
+{
+    ASSERT(document.isFullyActive());
+
+    Ref<CloseWatcher> watcher = adoptRef(*new CloseWatcher(document));
```
Nit: CTAD
```suggestion
    Ref watcher = adoptRef(*new CloseWatcher(document));
```

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcher.cpp:
```diff
@@ -0,0 +1,149 @@
+
+void CloseWatcher::requestClose()
+{
+    requestToClose();
+}
+
+bool CloseWatcher::requestToClose()
```
I am not following along, is there a reason we put the implementation inside `requestToClose()`?

lukewarlow replied:
Because the web exposed method has a return type of void but the internal algorithm has a return type of bool. If I just try to have the one method which returns bool the build fails.

```
JSDOMConvertBase.h:148:42: error: no viable conversion from
      'bool' to 'std::monostate'
  148 |         return JSConverter<IDL>::convert(std::forward<U>(value));
  ```

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcher.h:
```diff
@@ -0,0 +1,80 @@
+
+#pragma once
+
+#include "AbortSignal.h"
```
I think you can remove this header include and forward decl `class AbortSignal;`?

lukewarlow replied:
Same as with CloseWatcher in CloseWatcherManager because it's used inside of RefPtr/Ref it needs the full type (build fails otherwise).

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcher.h:
```diff
@@ -0,0 +1,80 @@
+    bool m_hasCancelEventListener { false };
+    bool m_hasCloseEventListener { false };
+    RefPtr<AbortSignal> m_signal;
+    uint32_t m_signalAlgorithm;
```
Nit: It's better to zero initialize this POD value rather than having it be indeterminate? `uint32_t m_signalAlgorithm {};`

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcher.h:
```diff
@@ -0,0 +1,80 @@
+
+    explicit CloseWatcher(Document&);
+
+    bool IsClosed() const { return m_closed; }
```
This method name feels off. It could do without the `Is` prefix, plus `isClosed()` would be more compliant with our casing guidelines.

lukewarlow replied:
Changed it to isActive and flipped the boolean to better match the spec. Happy to rename it if you feel there's a better name?

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcher.cpp:
```diff
@@ -0,0 +1,149 @@
+{
+    RefPtr document = dynamicDowncast<Document>(scriptExecutionContext());
+    if (IsClosed() || m_isRunningCancelAction || !document || !document->isFullyActive())
+        return;
```
We may want to make a small helper method for this check since we do this in two different sites.

aprotyas commented on Source/WebCore/html/closewatcher/CloseWatcherManager.h:
```diff
@@ -0,0 +1,49 @@
+
+#pragma once
+
+#include "CloseWatcher.h"
```
I believe you can get away with a forward declaration of `CloseWatcher`.

lukewarlow replied:
```
error: member access into incomplete type 'WebCore::CloseWatcher'
   62 |             ptr->deref();
   ```
   
I think because it's used inside of Ref<> it needs the full type. It fails to build if it's just a forward declaration.

aprotyas reviewed (commented):
Looks good! Mostly stylistic feedback

nt1m commented on Source/WebCore/html/closewatcher/CloseWatcherManager.cpp:
```diff
@@ -0,0 +1,92 @@
+        auto& group = m_groups.last();
+        Vector<Ref<CloseWatcher>> groupCopy(group);
+        for (auto it = groupCopy.rbegin(); it != groupCopy.rend(); ++it) {
+            Ref<CloseWatcher> watcher = *it;
```
```suggestion
        Vector<Ref<CloseWatcher>> groupCopy(group);
        for (Ref watcher : makeReversedRange(groupCopy)) {
```

nt1m commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -3932,6 +3932,10 @@ imported/w3c/web-platform-tests/feature-policy/feature-policy-frame-policy-allow
 imported/w3c/web-platform-tests/feature-policy/feature-policy-frame-policy-allowed-for-some.https.sub.html [ DumpJSConsoleLogInStdErr Failure Pass ]
 imported/w3c/web-platform-tests/feature-policy/feature-policy-frame-policy-disallowed-for-all.https.sub.html [ DumpJSConsoleLogInStdErr Failure Pass ]
 
+# Close Watcher tests rely on sending escape keys from test_driver which doesn't work on iOS webkit.org webkit.org/b/274832
```
webkit.org appears twice (we typically put the bug number in front of the test paths too)

aprotyas commented on LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nnn-CloseWatcher-dialog-popover-expected.txt:
```diff
@@ -1,4 +1,4 @@
 button
 
-FAIL Create a CloseWatcher without user activation; create a dialog without user activation; create a popover without user activation promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: CloseWatcher"
+FAIL Create a CloseWatcher without user activation; create a dialog without user activation; create a popover without user activation assert_array_equals: lengths differ, expected array ["dialog cancel[cancelable=false]", "CloseWatcher cancel[cancelable=false]", "CloseWatcher close", "dialog close"] length 4, got ["CloseWatcher cancel[cancelable=false]", "CloseWatcher close", "dialog cancel[cance ...
```
I suppose this failure is still expected.

lukewarlow replied:
Yeah follow up patches are needed to fully integrate with the popover and dialog elements. Once that's done these tests should all pass correctly.

aprotyas reviewed (commented):
Apologies for the delay with the follow-up review 🙇 This looks good to me

lukewarlow committed:
Implement the Close Watcher API
https://bugs.webkit.org/show_bug.cgi?id=263305

Reviewed by Abrar Rahman Protyasha.

Implement the Close Watcher API behind a feature flag.

This patch doesn't include the integration with Popovers or Dialogs.

Spec: https://html.spec.whatwg.org/#close-requests-and-close-watchers

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/abortsignal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/keydown-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/keypress-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/keyup-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/not-user-activation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/esc-key/synthetic-keyboard-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/event-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/frame-removal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/inside-event-listeners-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n-activate-preventDefault_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n-activate_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n-closerequest-n_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n-destroy-n_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/n_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nn-activate-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nn-dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nnn-CloseWatcher-dialog-popover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nnn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/ny-activate-preventDefault_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/ny_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyn_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyn_dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nynn-destroy_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nynn-destroy_dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nynn_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nynn_dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyyn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/nyyyn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/y_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yn-activate_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yn_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/ynn-CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yy_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yyn_CloseWatcher-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yyy-CloseWatcher-dialog-popover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yyy-activate-CloseWatcher-dialog-popover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/close-watcher/user-activation/yyy_CloseWatcher-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/close-watcher/abortsignal-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/close-watcher/basic-expected.txt: Added.
* LayoutTests/platform/win/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/html/closewatcher/CloseWatcher.cpp: Added.
(WebCore::CloseWatcher::create):
(WebCore::CloseWatcher::establish):
(WebCore::CloseWatcher::CloseWatcher):
(WebCore::CloseWatcher::requestClose):
(WebCore::CloseWatcher::requestToClose):
(WebCore::CloseWatcher::close):
(WebCore::CloseWatcher::canBeClosed const):
(WebCore::CloseWatcher::destroy):
(WebCore::CloseWatcher::eventListenersDidChange):
(WebCore::CloseWatcher::virtualHasPendingActivity const):
* Source/WebCore/html/closewatcher/CloseWatcher.h: Added.
* Source/WebCore/html/closewatcher/CloseWatcher.idl: Added.
* Source/WebCore/html/closewatcher/CloseWatcherManager.cpp: Added.
(WebCore::CloseWatcherManager::add):
(WebCore::CloseWatcherManager::remove):
(WebCore::CloseWatcherManager::notifyAboutUserActivation):
(WebCore::CloseWatcherManager::canPreventClose):
(WebCore::CloseWatcherManager::escapeKeyHandler):
* Source/WebCore/html/closewatcher/CloseWatcherManager.h: Added.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::defaultKeyboardEventHandler):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::notifyActivated):
(WebCore::LocalDOMWindow::closeWatcherManager):
* Source/WebCore/page/LocalDOMWindow.h:

Canonical link: https://commits.webkit.org/289529@main
